app/components/step-slider.jsx

```diff
@@ -0,0 +1,261 @@
+  useEffect(() => {
+    window.addEventListener('resize', resizeHandler)
+    return () => window.removeEventListener('resize', resizeHandler)
+  }, [])
+
+  // if the other props have changed, we need to rerender the children
+  // _this.otherProps is only changed if it's shallow - different
+  // _this is written directcly because we don't want to cause another rerender - we just want to save the value for next time
+  if (!shallowEqual(_this.otherProps, otherProps)) _this.otherProps = otherProps
+
+  // has to be useLaoutEffect not useEffect or transitions will get enabled before the first render of the children and it will be blurry
+  if (typeof window !== 'undefined') {
+    useLayoutEffect(() => {
+      if (navRef.current) {
+        let navRect = navRef.current.getBoundingClientRect()
+        if (navRect.height && navRect.width) setNavBarRect(navRect)
+      }
+      if (footerRef.current) {
+        let footRect = footerRef.current.getBoundingClientRect()
+        if (footRect.height && footRect.width) setFooterRect(footRect)
+      }
+    }, [navRef.current, footerRef.current])
+  }
+  if (typeof window !== 'undefined') useLayoutEffect(resizeHandler, [outerRef.current])
+
+  function reducer(state, action) {
+    switch (action.type) {
+      case 'increment':
+        // Set next navigated-to panel as seen so we can render it
+        let currentStep = Math.min(state.currentStep + 1, children.length - 1)
+
+        if (steps) {
+          let newStatuses = {
+            ...state.stepStatuses,
+            [currentStep]: { ...state.stepStatuses[currentStep], seen: true },
+          }
+          state.stepStatuses = Object.keys(newStatuses).map(key => newStatuses[key])
+        }
+
+        return {
+          ...state,
+          currentStep: currentStep,
+          sendDoneToParent: state.currentStep >= children.length - 1,
+        }
+      case 'decrement':
+        return {
+          ...state,
+          currentStep: Math.max(0, state.currentStep - 1),
+          sendDoneToParent: state.currentStep === children.length - 1,
+        }
+      case 'clearSendDoneToParent':
+        return { ...state, sendDoneToParent: false }
+    }
+  }
+  // Keep track of each step's seen/completion status
+  // Populate statuses with initial values
+  if (steps) steps[0].seen = true
+  const [state, dispatch] = useReducer(reducer, { currentStep: 0, sendDoneToParent: false, stepStatuses: steps })
+
+  // the children need to be cloned to have the onDone function applied, but we don't want to redo this every time we re-render
+  // so it's done in a memo
+  const clonedChildren = useMemo(
+    () =>
+      // Only render if component has been seen
+      children.map((child, index) =>
+        React.cloneElement(child, {
+          ...otherProps,
+          ...child.props,
+          key: [index],
+          onDone: valid => {
+            let newStatuses
+
+            if (steps) {
+              if (valid) {
```

Comment:
create a new action type in reducer for updateStatuses the call to that will update the statuses (which will be saved) and cause a rerender.

put most of this code in that action.